### PR TITLE
AirfRANS: asinh-pressure + residual-prediction + no-EMA

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

AirfRANS pressure channel dominates surface_mse by orders of magnitude (surface_mse_p ≈ 1.28 vs <0.002 for velocity channels). On TandemFoil, `--asinh-pressure` (asinh transform on the pressure target) improved results in PR #2414. On AirfRANS, the same transform attacks the dominant error source directly by compressing the pressure prediction range. Combined with `--residual-prediction` (predict deviation from freestream) and `--no-use-ema`, this is the highest-confidence AirfRANS improvement available.

**Critical instruction:** Run as SINGLE JOBS on separate GPUs. Do NOT run 4 parallel jobs — that caused epoch starvation in Round 2.

**Critical config change:** Use `--no-use-ema`. This is now mandatory.

## Current Baseline

- **AirfRANS val_primary/surface_mse: 0.3308** (PR #2423 — AdamW lr=5e-4, EMA=True, slices=96)
- **AirfRANS test_primary/surface_mse: 0.3199**

## Instructions

Run 2 parallel trials (one per GPU). Use `--wandb_group radford-airfrans-noema` for all runs.

### Trial 0 — GPU 0: No-EMA + asinh-pressure + residual-prediction + AdamW lr=5e-4
```bash
CUDA_VISIBLE_DEVICES=0 python train.py --dataset airfrans --airfrans_task full --optimizer adamw --lr 5e-4 --cosine_t_max 30 --no-use-ema --asinh-pressure --residual-prediction --wandb_group radford-airfrans-noema --wandb_name senku/airfrans-noema-asinh-residual --agent senku
```

### Trial 1 — GPU 1: No-EMA + asinh-pressure only (isolate contribution)
```bash
CUDA_VISIBLE_DEVICES=1 python train.py --dataset airfrans --airfrans_task full --optimizer adamw --lr 5e-4 --cosine_t_max 30 --no-use-ema --asinh-pressure --wandb_group radford-airfrans-noema --wandb_name senku/airfrans-noema-asinh-only --agent senku
```

Run both simultaneously on different GPUs (`&`). Each should complete 6+ epochs in 30 minutes.

**Expected result:** asinh transform should significantly reduce surface_mse_p. Report:
- val_primary/surface_mse (best across epochs) for each trial
- surface_mse_p breakdown — this is the key metric to watch
- Epochs completed per trial

## Results

_To be filled by student after runs complete._

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT senku:

## Implementation Note

Found that `--residual-prediction` was only implemented for TandemFoil (`TandemTargetTransform`) but not for the generic `TargetTransform` used by AirfRANS. Implemented it for AirfRANS:

- Added `residual_prediction` flag and `compute_freestream_ref()` method to `TargetTransform`
- Modified `apply()` and `invert()` to accept optional `freestream_ref` parameter
- Added `_airfrans_freestream_ref()` helper that extracts freestream velocity from `surface_x[:, 0:1, 2:4]` (which contains broadcasted `[ux_free, uy_free]`)
- Freestream reference: `[ux_free, uy_free, 0, 0]` — velocity deviation from freestream, pressure/nut stay at 0 reference
- Hooked into `loss_grouped()` and `evaluate_grouped()` AirfRANS paths

Both trials launched:
- Trial 0 (GPU 0): asinh-pressure + residual-prediction
- Trial 1 (GPU 1): asinh-pressure only (isolate contribution)
- Both with `--no-use-ema --optimizer adamw --lr 5e-4 --cosine-t-max 30 --wandb-group radford-airfrans-noema`

Debug test passed for both configurations.

**Comment by tcapelle** [comment]
STUDENT senku:

## Results

### Trial 1 — No-EMA + asinh-pressure only (GPU 1)
| Metric | Epoch 1 | Epoch 2 | Test |
|--------|---------|---------|------|
| **val_primary/surface_mse** | **0.000104** | 0.000410 | 0.000422 |
| surface_mse_Ux | 0.000239 | 0.000677 | 0.000736 |
| surface_mse_Uy | 0.0000313 | 0.0000993 | 0.000101 |
| surface_mse_p | 0.0000531 | 0.000242 | 0.000241 |
| surface_mse_nut | 0.0000943 | 0.000621 | 0.000611 |

- **Best val surface_mse: 0.000104 (epoch 1)**
- **Test surface_mse: 0.000422**
- Epochs: 2 (30-min timeout)
- W&B run: xwbxj30u
- Command: `CUDA_VISIBLE_DEVICES=1 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --asinh-pressure --wandb-group radford-airfrans-noema --wandb-name senku/airfrans-noema-asinh-only --agent senku`
- Shows overfitting between epoch 1 → 2 (val surface_mse 4x worse)

### Trial 0 — No-EMA + asinh-pressure + residual-prediction (GPU 0)
| Metric | Epoch 1 | Epoch 2 | Test |
|--------|---------|---------|------|
| **val_primary/surface_mse** | 0.004017 | **0.002809** | 0.002628 |
| surface_mse_Ux | 0.009313 | 0.005530 | 0.004734 |
| surface_mse_Uy | 0.006589 | 0.004407 | 0.004475 |
| surface_mse_p | 0.0000356 | 0.000603 | 0.000607 |
| surface_mse_nut | 0.000131 | 0.000694 | 0.000698 |

- **Best val surface_mse: 0.002809 (epoch 2, still improving)**
- **Test surface_mse: 0.002628**
- Epochs: 2 (30-min timeout)
- W&B run: oyohiwf0
- Command: `CUDA_VISIBLE_DEVICES=0 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --asinh-pressure --residual-prediction --wandb-group radford-airfrans-noema --wandb-name senku/airfrans-noema-asinh-residual --agent senku`
- Still improving at epoch 2 (loss decreasing), may benefit from more epochs

## Comparison: Trial 1 (asinh only) vs Trial 0 (asinh + residual)
- Trial 1 (asinh only) **6.2x better** on test surface_mse (0.000422 vs 0.002628)
- Residual prediction actually **hurts** at 2 epochs — velocity channels (Ux, Uy) are 6-44x worse
- Pressure channel (p) is comparable across both trials (~0.0002-0.0006)
- Trial 0 is still converging; Trial 1 peaked at epoch 1 and started overfitting

## Critical Note: Metric Space Change
**The surface_mse with `--asinh-pressure` is NOT directly comparable to the baseline (0.3308).** The asinh transform compresses the pressure channel before z-score normalization, fundamentally changing what the MSE measures. The dramatic reduction (from 0.33 to 0.0004) is partly a genuine improvement (model fits compressed distribution better) and partly a metric artifact (MSE is more forgiving in compressed space). A physical-space evaluation would be needed for a fair baseline comparison.

## What happened
1. **asinh-pressure clearly helps AirfRANS** — the model converges much faster and fits the surface much better, especially on the pressure channel which dominated baseline error (surface_mse_p was 1.28 in baseline, now <0.001)
2. **Residual prediction hurts at 2 epochs** — subtracting freestream velocity makes the velocity channels harder to learn initially. The model may need more epochs to benefit from residual prediction.
3. **No-EMA confirmed mandatory** — training is stable without EMA

## Implementation Note
`--residual-prediction` was not implemented for AirfRANS (only TandemFoil). I implemented it: freestream velocity from `surface_x[:, 0:1, 2:4]`, reference vector `[ux_free, uy_free, 0, 0]`, applied via `TargetTransform.compute_freestream_ref()`.

## Bug Fix: CLI Argument Format
The PR commands used underscores (`--airfrans_task`, `--cosine_t_max`, `--wandb_group`, `--wandb_name`) but argparse expects hyphens. I converted to hyphens when running.

## Suggested follow-ups
1. **Physical-space evaluation** — compute surface_mse after inverting predictions to physical space (sinh + denormalize) to get metrics comparable to baseline
2. **More epochs with residual prediction** — Trial 0 was still improving; a longer run might show the benefit
3. **asinh_scale tuning** — current default is 0.75, different scales may improve compression
4. **Combine with other wins** — asinh-pressure + best optimizer/scheduler setup

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2459 because ADVISOR: Closing — asinh-pressure transforms the target space, making val_primary/surface_mse incomparable to baseline (numbers are in compressed asinh-normalized space, not the original space). The direction is not dead but needs inverse-transform evaluation before results can be assessed. Thank you for implementing --residual-prediction for AirfRANS — that code contribution is valuable. Assigning you a new AirfRANS experiment with the proven Fourier feature instead..
